lib/lists.nix

```
@@ -86,15 +86,67 @@ rec {
         else op (foldl' (n - 1)) (elemAt list n);
     in foldl' (length list - 1);
 
-  /* Strict version of `foldl`.
+  /*
+    Reduce a list by applying a binary operator from left to right,
+    starting with an initial accumulator.
+    Before each application of the operator, the accumulator value is evaluated.
+    This behavior makes this function stricter than [`foldl`](#function-library-lib.lists.foldl).
 
-     The difference is that evaluation is forced upon access. Usually used
-     with small whole results (in contrast with lazily-generated list or large
-     lists where only a part is consumed.)
+    This strictness also allows avoiding an accumulation of thunks,
+    which would otherwise cause a stack overflow for large enough lists.
 
-     Type: foldl' :: (b -> a -> b) -> b -> [a] -> b
-  */
-  foldl' = builtins.foldl' or foldl;
+    Unlike [`builtins.foldl'`](https://nixos.org/manual/nix/unstable/language/builtins.html#builtins-foldl'),
+    the initial accumulator argument is evaluated before the first iteration.
+
+    A call like
+
+    ```nix
+    foldl' op acc₀ [ x₀ x₁ x₂ ... xₙ₋₁ xₙ ]
+    ```
+
+    is semantically the same as
```

```suggestion
    is denotationally equivalent to
```

If we're going to talk about the semantics, we have to pick the right semantics.

We should then also (later) discuss the other semantics, the operational semantics:

```diff
+However, operationally, the accₙ bindings are optimized away using a single memory location instead.
```

---

I actually don't think we need to mention the operational semantics. It's not very useful information and pretty much an implementation detail.

I am now using "is (denotationally) equivalent to" though.